server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java

```diff
@@ -166,6 +183,74 @@ public void testForceMergeCausesFullSnapshot() throws Exception {
         assertThat(secondSnapshotShardStatus.getIncrementalFileCount(), greaterThan(0));
     }
 
+    public void testRecordCorrectSegmentCountsWithBackgroundMerges() throws Exception {
+        final String repoName = "test-repo";
+        createRepository(repoName, "fs");
+
+        final String indexName = "test";
+        // disable merges
+        assertAcked(prepareCreate(indexName).setSettings(indexSettingsNoReplicas(1).put(MergePolicyConfig.INDEX_MERGE_ENABLED, "false")));
+
+        // create an empty snapshot so that later snapshots run as quickly as possible
+        createFullSnapshot(repoName, "empty");
+
+        // create a situation where we temporarily have a bunch of segments until the merges can catch up
+        long id = 0;
+        final int rounds = scaledRandomIntBetween(50, 300);
+        for (int i = 0; i < rounds; ++i) {
+            final int numDocs = scaledRandomIntBetween(100, 1000);
+            BulkRequestBuilder request = client().prepareBulk();
```

Comment:
I think it'd be better to refresh on each bulk to force a segment to be written:

```suggestion
            BulkRequestBuilder request = client().prepareBulk().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
```

I suspect if we did that we wouldn't need so many rounds.

Reply:
++ right done in d629c74